File: apps/web/lib/job-state.ts
export const jobStatuses = ["queued", "rendering", "complete", "failed"] as const;
export type JobStatus = (typeof jobStatuses)[number];

const allowedTransitions: Record<JobStatus, readonly JobStatus[]> = {
  queued: ["rendering", "failed"],
  rendering: ["complete", "failed", "queued"],
  complete: [],
  failed: ["queued"],
};

export function canTransitionJob(from: JobStatus, to: JobStatus): boolean {
  return allowedTransitions[from].includes(to);
}

export function assertJobTransition(from: JobStatus, to: JobStatus): void {
  if (!canTransitionJob(from, to)) {
    throw new Error(`Invalid render job transition: ${from} -> ${to}`);
  }
}

File: apps/worker/product_twin/job_state.py
from enum import StrEnum
class JobStatus(StrEnum): QUEUED='queued'; RENDERING='rendering'; COMPLETE='complete'; FAILED='failed'
ALLOWED={JobStatus.QUEUED:{JobStatus.RENDERING,JobStatus.FAILED},JobStatus.RENDERING:{JobStatus.COMPLETE,JobStatus.FAILED},JobStatus.COMPLETE:set(),JobStatus.FAILED:{JobStatus.QUEUED}}
def assert_transition(old:JobStatus,new:JobStatus)->None:
    if new not in ALLOWED[old]: raise ValueError(f'invalid job transition: {old} -> {new}')
